### ISSUE TITLE
Files locked in cache directory

### ISSUE THREAD

**Issue by seydu** [docker/for-win#525]
### Expected behavior
Process should have write permission to a cache directory and all its sub-directories)

### Actual behavior
When I start a container to run PHP FPM to serve Symfony2 some files in the cache directory are locked and the process cannot write in them. No way to remove remove the files (inside the container, in Windows using file explorer or Powershell, even when using Administrator privilege). When I stop Docker then I can remove the locked files.
For those familiar with Symfony2, this happens when Doctrine generates entity proxy class in temporary file and tries to rename it.

### Information
 -  Diagnostic ID: 1979D450-9549-4EDF-B258-1147D0CFAA65/2017-02-23_12-35-23
  - host distribution and version: OS: Windows 10 Enterprise Id: 1511 Build: 10586

### Steps to reproduce the behavior

  1. ...
  2. ...

**Comment by dgageot** [comment]
@seydu Would you have a compose file we could use to reproduce the issue?

**Comment by Toyinster** [comment]
+1 I have tried several ways to solve this but to no avail. I believe it is permission related: 
I have added `RUN usermod -u 1000 www-data` to the dockerfile, and manually set the permissions on the storage folder inside the container but neither of these measures have solved it. Any help here would be much appreciated.

This is the php error message:
`"rename(/opt/app/storage/proxies/__CG__DomainEntitiesOrganisation.php.58b29056532943.69863232,/opt/app/storage/proxies/__CG__DomainEntitiesOrganisation.php): No such file or directory"`

**Comment by Hrsher** [comment]
Same issue here - locked files - not only with php. Unlocks them only after Docker Settings -> Reset.
Can also cause very strage behavior in combination with git .. e.g. unable to checkout files, rm, reset etc.

I`m also unable to do chmod or chown on any mounted volume files - no effect.
This can be very frustrating for some setups.

Windows 10 Enterprise
Microsoft Windows [Version 10.0.14393]

Docker Community Edition
Version 17.03.1-ce-win5 (10743)
Channel: stable
b18e2a5

**Comment by chucky2305** [comment]
My colleague and i have the same problem.

e.g:
`/var/cache/dev/twig/19/199fa6689b84659fb14bddcbeec7d0ea05230a4bf173fdf13fb488dc4b5642db.php
`

This file cannot be deleted. Wheter from Windows, nor inside any container (with root)

Windows 10 Pro
Docker Community Edition
Version 17.03.1-ce-win5 (10743)
Channel: stable
b18e2a5

**Comment by strattas** [comment]
Same Here. It isn't super predictable or replicateable yet. But when building Magento2 sites in my local we have to often clean the pub directory and eventually each day the files will get locked and I have to restart docker to free them so I can delete and  continue my work. 
Windows 10 Pro 
Docker w Moby VM Version 17.03.1-ce-win12 (12058)

**Comment by rpander93** [comment]
Having exactly the same problem here. Problem occurs randomly to files generated by Doctrine, which are re-created on each request.

**Comment by jakubdrabik** [comment]
I have the same problem. Even after the container is killed, I must reset whole Docker to unlock the file.

Windows 10 Pro + the latest Docker version.

**Comment by TheRealDwright** [comment]
Also have this problem - npm creating locked files on the Windows Host. Can this be worked around with Docker Toolbox in the meantime?

**Comment by mrAndersen** [comment]
I have this issue too, in my case symfony can't rename or delete old doctrine cache fiel located in /var/www/symfony/var/cache/dev/doctrine/orm/Proxies/  locking all the project, only reboot helps

**Comment by Rainbowlemon** [comment]
I'm getting the same issue with the same setup. Any ETA on a possible fix? I've come to a dead end at the project I'm working on because the ProxyGenerator is producing files that can't be changed or removed. Here's my logs:

````
Warning: rename(/var/www/project/app/cache/dev/doctrine/orm/Proxies/__CG__AppBundleEntityAddressCurrent.php.5975ffd4a9d748.71864046,/var/www/project/app/cache/dev/doctrine/orm/Proxies/__CG__AppBundleEntityAddressCurrent.php): No such file or directory

1. in vendor/doctrine/common/lib/Doctrine/Common/Proxy/ProxyGenerator.php at line 306 

304        file_put_contents($tmpFileName, $proxyCode);
305        @chmod($tmpFileName, 0664);
306        rename($tmpFileName, $fileName);
307    }

**Comment by KevinBon** [comment]
Same here.

The issue show up when concurrent requests are called.

I run Docker for Windows with administrator privilege.

`Windows 10 Pro`
`Docker Community Edition Version 17.07.0-ce-rc1-win21 (12927)`

**Comment by SlvrEagle23** [comment]
Same exact problem here. Have verified that no other processes (backup tasks, etc) are accessing the files in the volume shared with the Docker container, and yet seemingly at random, files are being locked in very strange ways on the Windows host OS.

I had more than one case where the file appeared to Windows to have no owner information and to be unmodifiable, and only removing the Docker containers resolved the issue. Some other cases, the file just cannot be written to, but no explanation is given. Each situation is resolved completely by fully shutting down and removing the Docker containers.

I never encountered similar issues using Docker Toolbox and Virtualbox.

**Comment by dpgover** [comment]
I (and a lot of my co-workers) have the same issue. We are also using doctrine.
What I could get from all the cases we have of the problem, is that when two or more requests run at the same time, doctrine tries to generate those proxy files. Somehow, those processes collide trying to remove the proxy or one tries to delete the file and the other to access it. It happens more consistently when the parallel requests are ajax (no so sure why).

Once the files get locked, we have 4 strategies. Sometimes one works, sometimes does another:
1. Stop and up the containers.
2. Restart docker.
3. Delete the HyperV VM and restart docker.
4. Restart the PC (sometimes this does not even work)

I don't know what info we can provide that helps solving this issue.
We are using Win10 Pro 64bits.
Docker Comunity Edition 17.x (we have had this issue since version 12 or 13 I think)

Cheers.

**Comment by jaroslavlibal** [comment]
Same issue here with Doctrine in docker (proxies folder).

Docker for Windows 17.09.0-ce-win33 (13620)
Windows 10 Pro 1703

When I close the D4W app then the locked file disappears.

**Comment by Toshik** [comment]
Same here with `npm install` - files in `.bin` folder becomes locked, only Docker engine restart helps.

Version 17.09.0-ce-win33 (13620)
Channel: stable
8c56a3b

**Comment by jaroslavlibal** [comment]
Appears mostly when I send 2+ concurent requests to the webserver (eg. AJAX calls).

**Comment by dpgover** [comment]
@dgageot What info is needed that we can provide for the issue to advance...

Thanks!

**Comment by friism** [comment]
I'm guessing this happens for files that are bind-mounted from your Windows host into Docker containers?

If that's the case, you can maybe find a way to have the problematic files not be mapped to the Windows host and instead just be in the container or in a named volume (which won't touch the Windows filesystem).

As we make the transition to LCOW, we'll get a new and improved bind-mounting implementation: https://blog.docker.com/2017/09/preview-linux-containers-on-windows/

As that solution matures, you should check the behavior there, and we can work with Microsoft to fix things that are broken.

**Comment by k20human** [comment]
Same problem here with Symfony 3 and concurrent calls (REST API)
Some proxy files from doctrine are locked and not writeable ...

After a reboot of Docker i can remove the files

**Comment by liyo** [comment]
Same issue with Doctrine Proxies getting locked. After restarting Docker for Windows files can be removed but the issue persists.

**Comment by zeromodule** [comment]
I have a same problem with Docker for Windows, php-fpm, Doctrine and Symfony 3.
Doctrine proxies in cache directory getting locked and remain locked until a Docker restart occurs.

**Comment by darinrandal** [comment]
I'm having this same problem. We have a web application with nginx and it seems the most noticeable when switching branches with git (since it changes some files that are locked). It definitely seems like the problem gets worse when I start refreshing the site more causing some concurrency. I can confirm it clears the locks if I do docker-compose down and docker-compose up -d.

My only solution right now is before switching branches, clear all locked files via: Computer Management > System Tools > Shared Folders > Open Files > right click - disconnect all open files. You'll see all the files that are locked on this page and can clear them all easily without having to do down/up.

I'm not sure where to go from here but this problem is affecting everyone that works on this project, and with all the comments on this issue it definitely can't be an isolated incident.

**Comment by marhub** [comment]
Same problem here. Can't generate doctrine proxies on docker runingo on windows 10.
One year have passed since OP added this ticket and still not resolved...

**Comment by dpgover** [comment]
I solved the issue using an internal directory created when building the container. I asume it can also be done using named volumes.
That works... if you don't need to have direct access to the generated files.
Since my problem was with Doctrine proxies, it works just fine leaving them inside the container.

**Comment by SeBsZ** [comment]
Same problem, usually when a long-running script is executing in my browser, and I try to do some Git actions (checkout, merge). The file will be in a weird state, readonly, can't delete, can't read. The only way is to completely exit Docker. On Windows.

**Comment by chris-cacciatore** [comment]
Same here on windows, we have a quite big script that rebuilds a lot of files, and sometimes it cannot delete a few of them, and my software just crashes. Even from windows, it is impossible to delete it, I have to reboot docker for windows and the file is then automatically deleted.

I can see that this issue has been open for more than a year now, is someone working on it? It is too much time-consuming for us, I seriously consider dropping docker and advising my team to go for another solution...

**Comment by darinrandal** [comment]
@chris-cacciatore @SeBsZ I've found a "solution" that will help mitigate the problem until something can be truly done to fix it. This only works on Windows: Open Computer Management, navigate to Shared Folders then Open Files, right click anywhere in the open files area and choose "Disconnect All Open Files." I don't remember where I ended up finding this but it has worked well and a much better option than restarting Docker/Windows entirely.

**Comment by chris-cacciatore** [comment]
@darinrandal thanks, I will try that!

**Comment by darinrandal** [comment]
@chris-cacciatore I also need to add that this only works on the latest version of Docker, the one that utilizes the tray icon (not the Docker Quickstart Toolbox terminal one). Not sure why, but they seem to handle the files differently. Maybe it's because the newer version actually requests permission to share the drive whereas the older one does not.

I have since switched to Mac and I'm happy to note I have no issues with Docker anymore, presumably due to Mac's unix-like nature. This is with the same docker containers that gave me plenty of headache on Windows.

**Comment by carlead** [comment]
Hi, I have the same problem on my side with files locked on Docker / Windows 10. Did someone find a solution ? Thanks.

**Comment by chris-cacciatore** [comment]
@carlead : see darinrandal's workaround (a few posts ago), it works well for me!

**Comment by carlead** [comment]
Yes I see, but we already use this, but we have locked files every times we refresh webpage, so it's complicated to use this method with Disconnect All Opens Files. I hope there will be a real fix for this problem (not sure it's possible).

**Comment by darinrandal** [comment]
I've found an actual solution, though it's not universal and you may have to do some digging around for your specific solution. We use nginx for our web server which seemed to have been the cause of the file locks due to some cache options in our nginx.conf. After removing the following configs and restarting nginx we haven't experienced the same issues, so it's worth checking out. You'll likely have to check whatever web server or proxies you use and disable any sort of caching with them.

`open_file_cache max=200000 inactive=20s;
open_file_cache_valid 30s;
open_file_cache_min_uses 2;
open_file_cache_errors on;
sendfile on;`

**Comment by carlead** [comment]
We use apache not nginx, maybe there is something similar with apache, thank you !

**Comment by aftabnaveed** [comment]
I was having the same issue https://github.com/docker/for-win/issues/1968 with my doctrine generated proxies, point the proxies to /tmp folder in the container has fixed the problem for now.

**Comment by cybertk** [comment]
@darinrandal's workaround works fine for me on Docker Version 18.03.1-ce-win65 (17513)

**Comment by ktallafus** [comment]
Sadly, doing that is not always possible. I wish they could fix this. It´s so annoying. 

Most of the time I have this problem with the generated autoload.php file from composer in the vendor folder. I´m not even doing a composer update or something like that. It just happens out of the blue after some time while working on the project. 
Then I have to shutdown docker completely and start again. After that it´s working again.

**Comment by SlvrEagle23** [comment]
Sad to report that Docker for Windows' latest stable release version is still "eating" files from time to time. In particular, it seems to "eat" files that are written more than once in quick succession, which is often the case for cache files. Once the files are "eaten", they become locked and even the owning user can't modify the file, and restarting Docker entirely is retired.

/remove-lifecycle stale

**Comment by nsepetys** [comment]
Still happening for me on docker v. 18.06.1-ce-win73 (19507) . Lost a few hours of work because of the locking. Odd part is that I tried to access the locked file on the ubuntu container that shared the mounted folders and while it did show the file as listed an attempt to change permissions (as root) resulted in "... No such file or directory". As other have reported, restarting the docker process seems to resolve the issue but this is somewhat of a show stopper when I'm losing hours of work arbitrarily.

**Comment by Scentyoo** [comment]
I had the problem with the locked files for a long time... I was able to see the locked files in Windows 10 under "Computer Management > Shared Folders > Open Files". I deactivated the php opcache and the problem was gone. Maybe this helps other users, too

**Comment by dpgover** [comment]
@j0nnybrav079
I solved the issue (as said in a comment from 9 months ago... I cannot link it from my phone), specifically for the doctrine proxies issue; you can create a proxy directory inside the container when building, and make a symlink into your project. Then, tell doctrine to use that symlink folder. The files will be created and handled inside the container instead of being handled by windows

**Comment by j0nnybrav079** [comment]
> @j0nnybrav079
> I solved the issue (as said in a comment from 9 months ago... I cannot link it from my phone), specifically for the doctrine proxies issue; you can create a proxy directory inside the container when building, and make a symlink into your project. Then, tell doctrine to use that symlink folder. The files will be created and handled inside the container instead of being handled by windows

hi @dpgover  
and thank you very much for your reply. I already read your solution yesterday and it works like a charm! 

The only question on my side still is: "will this ever be fixed?", what, if i use some other packages, how can i know if they won't throw the same error or not,someday? Do you have any information @docker-for-desktop-robot  ? :)

**Comment by digitalstraw** [comment]
My solution in Doctrine was to use APCu as a cache for proxies. Doctrine became 400% faster on my Docker, by the way.

See https://github.com/merorafael/docker-php/blob/master/7.1/Dockerfile
See https://www.doctrine-project.org/projects/doctrine-orm/en/2.6/reference/advanced-configuration.html

**Comment by benniebotha** [comment]
This is still an issue. I hope they fix it soon. 
/remove-lifecycle stale

**Comment by onigunn** [comment]
Encountering this problems in a php+nginx enviroment too. 

/remove-lifecycle stale

**Comment by SlvrEagle23** [comment]
To share some additional information on exactly what's causing this issue on my end:

- Inside a Docker container, a cron task is reading and executing PHP files to perform routine synchronization tasks;
- The PHP files being used are bind-mounted to the Windows host, because they're actively being developed; and
- At the same time as the container is trying to read a given PHP file, something out on the host computer is also trying to access that file (either Windows trying to index it, backup software trying to mess with it, or an IDE trying to save changes to it).

This causes the given conflicted file to enter a state of existential crisis, where it's not modifiable by any user on the host OS. I believe, but I'm not 100% confident, that at the same time, the guest container considers that file handle "stale".

Shutting down the Docker Desktop engine (and thus the Hyper-V VM) fixes the file's state and allows the host OS to modify it again.

I have no idea what the underlying cause of this is, but I can reproduce it like clockwork on my end just by directly editing a file that the cron task is also reading at the same time, so if there's any debugging I can do, I'm happy to do so. This issue has plagued me for years and resulted in me going through a (very expensive) misadventure trying to find alternatives to Docker Desktop.

**Comment by blueamerican** [comment]
still an issue

**Comment by zpvini** [comment]
Running docker as admin fixed the doctrine issue for me.

**Comment by digitalstraw** [comment]
In Doctrine also helps to set `autoGenerateProxyClasses` to `false`. It is not necessary to have this feature on in local development on Windows, where this bug appears.

**Comment by blueamerican** [comment]
> 
> 
> When the files were locked, I managed to unlock them in Windows > Computer Management > Shared Folders > Open Files.
> 
> Running the docker app as admin fixed the doctrine issue for me.

How do you run docker (desktop) as an admin?

**Comment by zpvini** [comment]
> > When the files were locked, I managed to unlock them in Windows > Computer Management > Shared Folders > Open Files.
> > Running the docker app as admin fixed the doctrine issue for me.
> 
> How do you run docker (desktop) as an admin?

Open the start menu, right click on Docker Desktop > More > Run as Administrator.

**Comment by blueamerican** [comment]
> Open the start menu, right click on Docker Desktop > More > Run as Administrator.
run docker as an admin doesn't solve the issue for me.

**Comment by andrhevictor** [comment]
Moving proxies files to /tmp (if /tmp is not binded to your computer) folder solves the problem.
Just set:
` 'proxy_dir' => '/tmp'`

**Comment by Lagily** [comment]
had the same issue, setting an absolute path made it work for us, e.g /var/www/html/doctrine_proxy/

**Comment by mkuehnel** [comment]
running docker as admin doesn't seem ti fix the issue
@andrhevictor, @Lagily in which file do I have to put this config?

**Comment by Lagily** [comment]
I set it in the Doctrine PHP code config @mkuehnel 

$config->setProxyDir('/path/to/myproject/lib/MyProject/Proxies');

https://www.doctrine-project.org/projects/doctrine-orm/en/2.6/reference/advanced-configuration.html

**Comment by digitalstraw** [comment]
> This is terrible. My macbook is having the battery replaced so I'm on my windows desktop for the first time in 2 years and Symfony was running extremely slow (30+ seconds per request), so I tried setting up docker. Spent half a day yesterday getting it running just to see slow speeds and now files that I can't remove from the Proxies folder without restarting my computer (the same issue as people above). I can't believe this hasn't been solved in 2 years..

You can just set `autoGenerateProxyClasses: false` in your Doctrine local development (non-production) configuration. It helps...

**Comment by digitalstraw** [comment]
> Hmmm...after a few refreshes I got this:
> 
> ```
> Compile Error: Doctrine\Common\Proxy\AbstractProxyFactory::getProxyDefinition(): Failed opening required '/var/www/[MY_PROJECT]/var/cache/dev/doctrine/orm/Proxies/__CG__AppEntity[MY_ENTITY].php' (include_path='.:/usr/local/lib/php')
> ```

Try to delete the cache and run again. If it does not help, consider installing [APC](https://pecl.php.net/package/APCu) extension and configure Doctrine like this:

```
metadataCache: apcu
queryCache: apcu
```

**Comment by davclark** [comment]
It seems this issue has veered a bit from the initial problem. So, I'm happy to open another issue.

In any case, I have recently had a couple different scenarios where lock files got permanently "stuck" unless docker for windows was stopped (even resetting didn't fix the issue as long as Docker was running). This happened twice with Git operations inside a container on a bind-mount, and once with my own lockfile (set from redis / rq) again on a bind-mounted directory.

Any pointers on how to avoid this would be greatly appreciated. We can try to change how we do the lock in RQ, and could change the version of git we're using perhaps?

**Comment by digitalstraw** [comment]
@davclark My files get stuck always when I run one script operating over one set of filesystem files in several parallel instances/requests. Files get stuck, then. Not all, some of them. So maybe if you avoid parallel access to one file, you might avoid this...

**Comment by spaceneenja** [comment]
I get similar issues when trying to npm install. Oddly this didn't appear until the past several weeks.

This seems part of to be a whole family of problems with LCOW that surface after some update or another and not really related to Docker Desktop.

**Comment by davclark** [comment]
Thanks @digitalstraw - though I am a bit loathe to patch both git and redis/rq to change their file locking behavior!

**Comment by davclark** [comment]
I saw that 2.2.3.0 had a bunch of changes to bind mounts. But I didn't see anything regarding this issue. Are there any updates in that 2.2.3 or the near future? The urgency of this issue goes up now that WSL2 is the default in 2.2.3 - which I think means that stable channel gets WSL2 as default in June.

**Comment by dthree** [comment]
Same exact issue but on OSX.

**Comment by Dream-117** [comment]
I accidentally copied folders and directories into docker container when I tried to import a .pkg file. Now my windows 10 download folder is not accessible and docker is also shut down. How do I resolve the issue of resetting both my windows drives and set docker too to original settings
